### DISCUSSION TITLE
How to tell if the host opens/closes CDC serial port

### DISCUSSION THREAD

**Discussion by whowechina** [hathach/tinyusb#2614] [Q&A] [answered]
Hi, I'm using CDC serial ports on rp2040.
It seems there's no way I can tell if the host opens or closes the port.
I tried tud_cdc_n_connected(), it only reports if the whole USB is connected.
I tried tud_cdc_n_get_line_coding(), I can tell the "open" from the baudrate change, but I can't tell a "close" as the baudrate never changes back.
I tried tud_cdc_n_get_line_state(), it always reports "0" on my system.

So I'm wondering, is it a mission impossible, or was I going a wrong direction?

**Comment by HiFiPhile** [comment, marked as answer]
I'm afraid it's impossible to detect it reliability at least on Windows, there is only one `SET_CONTROL_LINE_STATE` request sent from host with both DTR and RTS equal to 0 on port closing.

If you have your own host software then you can set DTR=1 at opening, then you can use DTR as closing detection.

**Comment by whowechina** [comment]
Thank you!
